controllers/messages.py

```diff
@@ -330,6 +356,110 @@ async def delete_thread(  # TODO: This should be renamed to delete_message
+        item_id=message_id,
+        item_type=Message,
+    )
+
+    # check if we are deleting the from or for message
+    if delete_item.for_id == token_payload["id"]:
+        delete_item.for_deleted = True
+    elif delete_item.from_id == token_payload["id"]:
+        delete_item.from_deleted = True
+    else:
+        # The user is attempting to delete another user's messages
+        raise AuthError(
+            {
+                "code": 403,
+                "description": "You do not have permission to "
+                "delete another user's messages.",
+            },
+            403,
+        )
+
+    # just mark the object for deletion - we can remove it properly via a separate
+    # offline data cleaning process later (probably needs for_deleted and from_deleted
+    # changed to dates rather than bools)
+    await sah_config.db.update_object(delete_item, current_user_id=token_payload["id"])
+
+    return jsonify({"success": True, "deleted": message_id})
+
+
+@messages_endpoints.route("/thread/<thread_id>", methods=["DELETE"])
+@requires_auth(sah_config, ["delete:messages"])
+async def delete_thread(
+    token_payload: UserData,
+    thread_id: int,
+) -> Response:
+    validator.check_type(thread_id, "Message ID")
+    thread_id = int(thread_id)  # flask typing is rubbish
+
+    delete_item = await sah_config.db.one_or_404(
+        item_id=thread_id,
+        item_type=Thread,
+    )
+
+    # Check if the user is attempting to delete another user's threads
+    if (
+        delete_item.user_1_id != token_payload["id"]
+        and delete_item.user_2_id != token_payload["id"]
+    ):
+        raise AuthError(
+            {
+                "code": 403,
+                "description": "You do not have permission to "
+                "delete another user's thread messages.",
+            },
+            403,
+        )
+
+    # For each message that wasn't deleted by the other user, the
+    # value of for_deleted/from_deleted (depending on which of the users
+    # it is) is updated to True
+    from_stmt = (
+        update(Message)
+        .where(
+            and_(
+                Message.thread == delete_item.id,
+                Message.for_id == token_payload["id"],
+                Message.from_deleted == false(),
+            )
+        )
+        .values(for_deleted=true())
+    )
+
+    for_stmt = (
+        update(Message)
+        .where(
+            and_(
+                Message.thread == delete_item.id,
+                Message.from_id == token_payload["id"],
+                Message.for_deleted == false(),
+            )
+        )
+        .values(from_deleted=true())
+    )
+
+    # just mark the objects for deletion - we can remove it properly via a separate
+    # offline data cleaning process later (probably needs for_deleted and from_deleted
+    # changed to dates rather than bools)
+    await sah_config.db.update_object(delete_item, current_user_id=token_payload["id"])
```

Comment:
That's actually not necessary, as the thread itself doesn't have any updatable properties. The `user_1_deleted` and `user_2_deleted` are computed, so we're not changing anything in the thread anymore